ADBDEV-6314: Fix unexpected output for empty grouping set

mos65o2 opened the pull request:
Fix unexpected output for empty grouping set

Problem:
Queries with grouping expressions and ungrouped attributes in TL could
produce unexpected output if the group did not have a primary key.

Example with 3 columns where "a" is a primary key:
     SELECT a,b,c FROM t GROUP BY GROUPING SETS((a),(b));

Grouping "(a)" had correct output of columns "a", "b", "c". Attribute "b" in
group "(b)" is not a primary key, so it is not unique, so it can be grouped.
In this case, the output of column "c" was undefined. This also applied to
"GROUPING SETS ((a),())", ROLLUP and CUBE (detailed behavior is
described in tests with ungrouped check).
Grouping expressions were introduced in PostgreSQL 9.5 ([f3d3118](https://github.com/postgres/postgres/commit/f3d3118532175541a9a96ed78881a3b04a057128)) and
such queries were prohibited. And this patch aligns the behavior with that.

Solution:
Improve parse-stage validation for ungrouped columns when using grouping
expressions.

Changes:
Similar to PostgreSQL, the parsing stage for queries with grouping
expressions reads common attributes for all groups. If any, they are placed in
"groupClauseCommonVars". In case of ungrouped attributes in TL, this list is
used to check for PK in groups.
Compared to the original, the "groupClauses" list uses TargetEntry rather than
an expression. Added function to get common refsortgroupref in grouping
clauses. At the beginning, the function assumes that all expressions are found
in grouping clauses. By running through the groupings, missing expressions
are excluded from the common refs list. The extraction function in the form
"list_intersection_int" taken from f3d3118.
The patch does not change the behavior of regular GROUP BY.

Co-authored-by: anarazel <andres@anarazel.de>

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/85622

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/85855

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/85904

mos65o2 committed:
Fix parser to check ungrouped attributes in TL

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/86054

mos65o2 commented on src/backend/parser/parse_agg.c:
```diff
@@ -840,35 +847,47 @@ parseCheckAggregates(ParseState *pstate, Query *qry)
 
 	/*
-	 * Build a list of the acceptable GROUP BY expressions for use by
+	 * Build a list of the acceptable unique GROUP BY expressions for use by
```
In fact, the TLEs on this list are not unique to some cases.

red1452 reviewed (commented):
You wrote, that changes are similar to Postgres 9.5, but I did not find the same sources there. Can you give the commit, which is example of your patch? Or can you backport such commit?

whitehawk commented:
In the description:
> "groping extensions"

groping -> grouping

And I guess not "extensions", but "expressions". Need to update it throughout the text.

Also, it is better to use past tense when describing previous wrong behaviour, and present tense for new/current behaviour.

whitehawk commented:
When I try the `functional_deps` solely (with optimizer enabled), it fails on my side.
Steps:

```
cd ./src/test/regress
PGOPTIONS="-c optimizer=on" ./pg_regress --init-file=init_file functional_deps
```

results in:
```diff
--- \/usr\/re\/share\/review\/ws58\-ADBDEV\-6314\/gpdb\/src\/test\/regress\/expected\/functional_deps_optimizer\.out	2024-11-19 14:24:19.752063949 +1000
+++ \/usr\/re\/share\/review\/ws58\-ADBDEV\-6314\/gpdb\/src\/test\/regress\/results\/functional_deps\.out	2024-11-19 14:24:19.771073451 +1000
@@ -149,16 +149,16 @@
 QUERY PLAN
 ___________
  GroupAggregate
-   Group Key: funcdep1.a, funcdep1.b, funcdep1.c
+   Group Key: t1.a, t1.b, t1.c
    ->  Sort
-         Sort Key: funcdep1.a, funcdep1.b, funcdep1.c
+         Sort Key: t1.a, t1.b, t1.c
          ->  Hash Join
-               Hash Cond: (funcdep1.b = funcdep2.b)
+               Hash Cond: (t1.b = t2.b)
                ->  Gather Motion 3:1  (slice1; segments: 3)
-                     ->  Seq Scan on funcdep1
+                     ->  Seq Scan on funcdep1 t1
                ->  Hash
                      ->  Gather Motion 3:1  (slice2; segments: 3)
-                           ->  Seq Scan on funcdep2
+                           ->  Seq Scan on funcdep2 t2
 GP_IGNORE:(12 rows)
 
 select sum(t2.a), t1.a, t1.b, t1.c from funcdep1 t1 join funcdep2 t2 on t1.b = t2.b group by t1.a;

======================================================================
```

It appears that it is due to GUC `optimizer_enable_table_alias`. If I set it to `off` in the test, the test passes.
For ex with sample patch below:
```diff
diff --git a/src/test/regress/sql/functional_deps.sql b/src/test/regress/sql/functional_deps.sql
index 878a595259..5026b275f9 100644
--- a/src/test/regress/sql/functional_deps.sql
+++ b/src/test/regress/sql/functional_deps.sql
@@ -2,6 +2,9 @@
 
 -- Enable logs that allow to show GPORCA failed to produce a plan
 SET optimizer_trace_fallback = on;
+-- start_ignore
+SET optimizer_enable_table_alias = off;
+-- end_ignore
 
 CREATE TEMP TABLE articles (
     id int CONSTRAINT articles_pkey PRIMARY KEY,
@@ -429,3 +432,6 @@ DROP TABLE test_table3;
 
 RESET optimizer_trace_fallback;
 RESET optimizer;
+-- start_ignore
+RESET optimizer_enable_table_alias;
+-- end_ignore

```
I guess that the CI pipeline passes Ok, because some other test disables this GUC, and doesn't reset it back.

It is a separate issue (as it was presented before your patch). But I believe we need to keep tests as low coupled with each other, as possible. So it is better to fix this problem some day soon. Can you please re-check on your side and create a ticket for this problem?

mos65o2 commented:
> You wrote, that changes are similar to Postgres 9.5, but I did not find the same sources there. Can you give the commit, which is example of your patch?
 
[f3d3118](https://github.com/postgres/postgres/commit/f3d3118532175541a9a96ed78881a3b04a057128) introduces support in PG 9.5. Also in [GPDB7](https://github.com/arenadata/gpdb/commit/f3d3118532175541a9a96ed78881a3b04a057128).
> Or can you backport such commit?
 
I discussed this issue with @Stolb27. I only need to correct the parser behavior without porting. My changes only concern the ungrouped columns check. Whereas the patch in PostgreSQL makes changes to the planner and much more.

red1452 commented:
> When I try the `functional_deps` solely (with optimizer enabled), it fails on my side. 
> It appears that it is due to GUC `optimizer_enable_table_alias`. If I set it to `off` in the test, the test passes. For ex with sample patch below:
> 
> ```diff
> diff --git a/src/test/regress/sql/functional_deps.sql b/src/test/regress/sql/functional_deps.sql
> index 878a595259..5026b275f9 100644
> --- a/src/test/regress/sql/functional_deps.sql
> +++ b/src/test/regress/sql/functional_deps.sql
> @@ -2,6 +2,9 @@
>  
>  -- Enable logs that allow to show GPORCA failed to produce a plan
>  SET optimizer_trace_fallback = on;
> +-- start_ignore
> +SET optimizer_enable_table_alias = off;
> +-- end_ignore
>  
>  CREATE TEMP TABLE articles (
>      id int CONSTRAINT articles_pkey PRIMARY KEY,
> @@ -429,3 +432,6 @@ DROP TABLE test_table3;
>  
>  RESET optimizer_trace_fallback;
>  RESET optimizer;
> +-- start_ignore
> +RESET optimizer_enable_table_alias;
> +-- end_ignore
> ```
> 
> I guess that the CI pipeline passes Ok, because some other test disables this GUC, and doesn't reset it back.
> 
> It is a separate issue (as it was presented before your patch). But I believe we need to keep tests as low coupled with each other, as possible. So it is better to fix this problem some day soon. Can you please re-check on your side and create a ticket for this problem?

Tests for GPDB6 is run with PGOPTIONS="optimizer_enable_table_alias=off".  You do not need to set GUC in the test.

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1031,9 +1064,12 @@ check_ungrouped_columns_walker(Node *node,
 	 */
 	if (context->have_non_var_grouping && context->sublevels_up == 0)
 	{
+
```
excessive line

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1544,145 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+static List*
+get_com_grouping_ressortgroupref_routine(Node *grpcl, List *targetList, List *com_refs)
+{
+
```
excessive line

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1544,145 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+		TargetEntry *tle = get_sortgroupclause_tle((SortGroupClause *) grpcl, targetList);
+		result = lappend_int(result, tle->ressortgroupref);
+	}
+
```
excessive line

mos65o2 replied:
I delete similar space in "get_groupclause_tles" too.

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1544,145 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+ */
+List*
+get_com_grouping_ressortgroupref(Query *qry, bool *hasGrouping, List *grp_tles){
+
```
excessive line

mos65o2 replied:
Removed this line

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1544,145 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+
+	foreach(l, qry->groupClause)
+	{
+
```
excessive line

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -894,14 +913,24 @@ parseCheckAggregates(ParseState *pstate, Query *qry)
 			have_non_var_grouping = true;
-			break;
+		}
+			else if (!hasGrouping ||
```
please fix indent here

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1544,145 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+
+	List *result = NIL;
+
+	if ( !grpcl)
```
excessive space

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1544,145 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+		GroupingClause *gc = (GroupingClause*)grpcl;
+
+		/* Cube and Rollup do not contain common attributes */
+		if(gc->groupType == GROUPINGTYPE_CUBE || gc->groupType == GROUPINGTYPE_ROLLUP){
```
Please move brace to the next line

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1544,145 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+					com_refs = list_delete_int(com_refs, lfirst_int(lc_del));
+				}
+				else
+					lc = lc->next;
```
Smaller branch is recommended to place at a first place in the 'if' statement.

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1544,145 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+			continue;
+
+		/* Scan in grouping extensions*/
+		if(IsA(grp, GroupingClause)){
```
Please move brace to the next line

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1544,145 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+				get_com_grouping_ressortgroupref_routine(grp, qry->targetList,com_grouping_expr_ressortgroupref);
+		}
+		/* Scan in group by*/
+		if (IsA(grp, SortGroupClause)){
```
Please move brace to the next line

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -814,12 +818,15 @@ parseCheckAggregates(ParseState *pstate, Query *qry)
 	List	   *groupClauses = NIL;
 	bool		have_non_var_grouping;
+	List	   *groupClauseCommonVars = NIL;
+	List 	   *com_grouping_ressortgroupref = NIL;
```
Please use consistent style. Here CamelCase and snake_case is mixed.

whitehawk commented:
In the description:
> in TL

I think 'TL' is not so common abbreviation. I suggest to write full wording here.

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -840,35 +847,47 @@ parseCheckAggregates(ParseState *pstate, Query *qry)
-	 * Build a list of the acceptable GROUP BY expressions for use by
+	 * Build a list of the acceptable unique GROUP BY expressions for use by
 	 * check_ungrouped_columns().
+	 *
```
Can't we replace the entire loop below with a walker similar to [CQueryMutators::GetGroupUniqueTleReferencesWalker](https://github.com/arenadata/gpdb/blob/ae07e294ffce0edd763529acc0a9aac4ec7f1f8b/src/backend/gpopt/translate/CQueryMutators.cpp#L179) ?

mos65o2 replied:
Yes, this is very similar to what we need. But in this cycle we need to get TLE, not refsortgroupref. Because in further processing either TLE or expr is used.

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1031,9 +1064,12 @@ check_ungrouped_columns_walker(Node *node,
-			if (equal(node, lfirst(gl)))
+			TargetEntry *tle = lfirst(gl);
+
+			  if (equal(node, tle->expr))
```
indent

whitehawk commented:
Followins query:
```
create table funcdep1(a int primary key, b int, c int, d int);
explain (costs off) select a, b, c from funcdep1 group by GROUPING SETS(a), GROUPING SETS(a, ());
```
On Postgres 17 works ok:
```
postgres=# select version();
create table funcdep1(a int primary key, b int, c int, d int);
explain (costs off) select a, b, c from funcdep1 group by GROUPING SETS(a), GROUPING SETS(a, ());
                                                 version                                                  
----------------------------------------------------------------------------------------------------------
 PostgreSQL 17devel on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0, 64-bit
(1 row)

CREATE TABLE
         QUERY PLAN         
----------------------------
 HashAggregate
   Hash Key: a
   Hash Key: a
   ->  Seq Scan on funcdep1
(4 rows)

```

while on your branch it reports an error:
```
postgres=# select version();
create table funcdep1(a int primary key, b int, c int, d int);
explain (costs off) select a, b, c from funcdep1 group by GROUPING SETS(a), GROUPING SETS(a, ());
                                                                                                                  version                                                                                                                  
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 9.4.26 (Greenplum Database 6.27.1_arenadata57+dev.51.gae07e294ff build dev) on x86_64-unknown-linux-gnu, compiled by gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0, 64-bit compiled on Nov 19 2024 11:50:41 (with assert checking)
(1 row)

CREATE TABLE
ERROR:  column "funcdep1.b" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: explain (costs off) select a, b, c from funcdep1 group by GR...
```

Can you please check?

mos65o2 commented:
> Followins query:
`GROUPING SETS(a), GROUPING SETS(a, ());`
On Postgres 17 works ok

Also runs in PG 9.5. Maybe I should make a list of common attributes for a specific grouping expression.

mos65o2 committed:
Code refactoring

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/86371

mos65o2 committed:
Fix serial GROUPING SET check

Also testcases are added

mos65o2 commented:
It is necessary to select common attributes among top-level groupings.
This case degenerates into a simple GROUP BY, despite the absence of common attributes among the top-level GROUPING SETS, so these queries work:
`create table t (a int primary key, b int, c int, d int);`

`select a, b, c from t group by GROUPING SETS(a), GROUPING SETS (a,());` ->
 `select a, b, c from t group by a, GROUPING SETS (a,());`

`select a, b, c from t group by GROUPING SETS(a), GROUPING SETS (b,()); `-> 
`select a, b, c from t group by a, GROUPING SETS (b,());`

It is necessary to take into account nesting. This query also degenerates into a regular GROUP BY and works:
`select a, b, c from t group by GROUPING SETS (GROUPING SETS ((a), GROUPING SETS ((grouping sets(a),a)),
			       GROUPING SETS (b,());`

This query does not work:
`select a, b, c from t group by GROUPING SETS (GROUPING SETS(a), GROUPING SETS (GROUPING SETS(b),a)),
			       GROUPING SETS (b,());`

With a composite key it works differently (does not work):
`create table t2 (a int, b text, c int, PRIMARY KEY(a,c));`
`select a, b, c from t2 group by GROUPING SETS (a,c), GROUPING SETS (a,())`
All queries are tested in PG 9.5.

mos65o2 commented:
> Followins query:
> 
> ```
> create table funcdep1(a int primary key, b int, c int, d int);
> explain (costs off) select a, b, c from funcdep1 group by GROUPING SETS(a), GROUPING SETS(a, ());
> ```
> while on your branch it reports an error:

I have changed the behavior. Please check. Pay attention to the tests.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/86406

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1538,135 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+ */
+List*
+get_com_grouping_ressortgroupref(Query *qry, List *grp_tles){
+	if(!qry || !grp_tles)
```
Add space between space and bracket.

mos65o2 replied:
Fixed

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1538,135 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+				if(list_member_int(grouping_refs, lfirst_int(lc)))
+					lc = lc->next;
+				else
+				{
```
If there is block with "{" - "}" you should use "{" and "}" and the other parts of such "if"

mos65o2 replied:
Fixed

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1538,135 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+			/*Exclude refs that did not appear in this grouping*/
+			ListCell *lc;
+			lc = list_head(com_refs);
+			while(lc){
```
Add space between while and bracket, and add space between ")" and "{".

mos65o2 replied:
Fixed

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1538,135 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+
+			/*Exclude refs that did not appear in this grouping*/
+			ListCell *lc;
+			lc = list_head(com_refs);
```
What about `ListCell *lc = list_head(com_refs);` ?

mos65o2 replied:
Fixed

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1507,14 +1538,135 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+				else
+				{
+					ListCell *lc_del = lc;
+					lc = lc->next;
```
lc = lnext(lc);

mos65o2 replied:
Fixed

red1452 reviewed (commented):
Add note about [f3d3118](https://github.com/postgres/postgres/commit/f3d3118532175541a9a96ed78881a3b04a057128) to the description.

mos65o2 committed:
Code refactoring

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/86860

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1536,138 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
 	return result;
 }
 
+static List*
```
Please add comments describing what this function does

mos65o2 replied:
Added

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1536,138 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+
+			/*Exclude refs that did not appear in this grouping*/
+			ListCell *lc = list_head(com_refs);
+			while (lc) {
```
```suggestion
			while (lc)
			{
```

mos65o2 replied:
Fixed

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1536,138 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+
+			/*Exclude refs that did not appear in this grouping*/
+			ListCell *lc = list_head(com_refs);
+			while (lc) {
```
Can't we use `foreach` to loop over the list?
```
#define foreach(cell, l)	\
	for ((cell) = list_head(l); (cell) != NULL; (cell) = lnext(cell))
```

mos65o2 replied:
`foreach` does not work correctly if you delete elements from iterated list. If you delete a cell, then the next iteration of `lnext(cell)` will call NULL ptr.

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1536,138 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+				{
+					ListCell *lc_del = lc;
+					lc = lnext(lc);
+					com_refs = list_delete_int(com_refs, lfirst_int(lc_del));					
```
there are several tab symbols at the end of this line

mos65o2 replied:
Fixed

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1536,138 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+	if (!qry || !grp_tles)
+		return NIL;
+
+	List *com_grouping_ressortgroupref = NIL;
```
Better to move the variable declarations to the start of the function.

mos65o2 replied:
Fixed

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1536,138 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+ * 	   clauses.
+ */
+List*
+get_com_grouping_ressortgroupref(Query *qry, List *grp_tles){
```
```suggestion
get_com_grouping_ressortgroupref(Query *qry, List *grp_tles)
{
```

mos65o2 replied:
Fixed

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1536,138 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+		if (IsA(grp, SortGroupClause))
+		{
+			TargetEntry *tle = get_sortgroupclause_tle((SortGroupClause *) grp, qry->targetList);
+			com_group_expr_ressortgroupref = lappend_int(com_group_expr_ressortgroupref, tle->ressortgroupref);
```
```suggestion
			com_grouping_ressortgroupref = list_append_unique_int(com_grouping_ressortgroupref, tle->ressortgroupref);
```

Thus
```
List *com_group_expr_ressortgroupref = NIL;
``` 
and 
```
/* Form a list of common refs for all clauses */
com_grouping_ressortgroupref = list_concat_unique_int(com_grouping_ressortgroupref, com_group_expr_ressortgroupref);
list_free(com_group_expr_ressortgroupref);
```
can be moved inside the block
```
if (IsA(grp, GroupingClause))
{
}
```

mos65o2 replied:
Fixed

mos65o2 committed:
Code refactoring

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/86899

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1536,153 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+				return NIL;
+
+			/*Exclude refs that did not appear in this grouping*/
+			ListCell *lc = list_head(com_refs);
+			while (lc)
+			{
+				if (list_member_int(grouping_refs, lfirst_int(lc)))
+				{
+					lc = lnext(lc);
+				}
+				else
+				{
+					ListCell *lc_del = lc;
+					lc = lnext(lc);
+					com_refs = list_delete_int(com_refs, lfirst_int(lc_del));
+				}
+			}
```
I suggest to backport `list_intersection_int()` from 'adb-7.2.0' and simpify this code 
```suggestion
			com_refs = list_intersection_int(com_refs, grouping_refs);
```

mos65o2 replied:
Done

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -842,33 +847,45 @@ parseCheckAggregates(ParseState *pstate, Query *qry)
 	}
 
+	/*
+	 * Getting adjacent TargetEntrys in groupings. Common tle will be needed
```
I doubt that the 'adjacent' is a good term here.
And what is a 'common ressortgroupref' should be explicitly explained in the comment.

mos65o2 replied:
Done. I hope the new description is exhaustive.

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1536,153 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+					list_append_unique_int(com_group_expr_ressortgroupref, te->ressortgroupref);
+			}
+			com_group_expr_ressortgroupref =
+				get_com_grouping_ressortgroupref_routine(grp, qry->targetList,com_group_expr_ressortgroupref);
```
```suggestion
				get_com_grouping_ressortgroupref_routine(grp, qry->targetList, com_group_expr_ressortgroupref);
```

mos65o2 replied:
Done.

mos65o2 committed:
Add list_intersection_int function

Modify get_com_grouping_ressortgroupref_routine.
Uses list_intersection_int instead of while cycle.
The function was taken from f3d3118.
Also code refactoring.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/87081

whitehawk commented on src/backend/parser/parse_agg.c:
```diff
@@ -878,11 +878,12 @@ parseCheckAggregates(ParseState *pstate, Query *qry)
-	 * after flatten_joinalias_vars. We also find out if there are grouping
-	 * expressions.
+	 * Getting ressortgrouprefs of common TargetEntrys in groupings.
+	 * TargetEntry is common if it appears in several grouping expressions
```
> "is common if it appears in several grouping expressions
	 * (including nesting) and in a regular GROUP BY (general case)."

Not if 
> "is common if it appears in all grouping expressions
	 * (including nesting) or in a regular GROUP BY (general case)."

?

mos65o2 replied:
Then for the case of `GROUPING SETS(a)` I will point out that it degenerates into a regular GROUP BY.

mos65o2 replied:
Done

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1537,140 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+{
+	List *result = NIL;
+
+	if ( !grpcl )
```
extra spaces: `if (!grpcl)`

mos65o2 replied:
Done

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1537,140 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+	List *result = NIL;
+
+	if ( !grpcl )
+		return result;
```
do `return NIL;`, doing `return result` is better only if it does not empty.

mos65o2 replied:
Done

red1452 commented on src/backend/nodes/list.c:
```diff
@@ -837,6 +837,32 @@ list_intersection(const List *list1, const List *list2)
+	Assert(IsIntegerList(list1));
+	Assert(IsIntegerList(list2));
+
+	result = NIL;
```
List	   *result = NIL;

mos65o2 replied:
Done

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1537,140 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+		foreach(lc, tles)
+		{
+			List *chunk = get_com_grouping_ressortgroupref_routine((Node *)lfirst(lc), targetList, com_refs);
+			if(!chunk)
```
space

mos65o2 replied:
Done

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1537,140 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+	{
+		TargetEntry *tle = get_sortgroupclause_tle((SortGroupClause *) grpcl, targetList);
+		result = lappend_int(result, tle->ressortgroupref);
+	}
```
If you add `return result;` here, you may remove the last `else` with brackets and one tab in each line inside.

mos65o2 replied:
Done. This solution seems less clear.

mos65o2 committed:
Fix comment com_ressortgroupref

mos65o2 committed:
Modifi if-else in ressortgroupref_routine

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/87195

red1452 commented on src/backend/parser/parse_agg.c:
```diff
@@ -1505,16 +1538,136 @@ get_groupclause_exprs(Node *grpcl, List *targetList)
+
+		return com_refs;
+	}
+	List *tles = (List *)grpcl;
```
Add one empty line after bracket.

mos65o2 replied:
Done.

mos65o2 committed:
Add empty line after cycle in routine

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/87490

whitehawk commented:
In the start of the description:
> Prohibiting grouping sets with ungrouped attributes in TargetList

Is it supposed to be a summary of the commit? If yes, it should be in Imperative mood and aligned with the PR description (so it should be "Fix unexpected output for empty grouping set" I think). 

If it is supposed to be a part of commit's detailed description, I think you should avoid using Present Participle in isolation. So it is better to rephrase to smth like "This patch prohibits grouping sets with ungrouped attributes in TargetList.".

I'll add my approve for this PR, assuming you'll fix this small issue in the description.

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-6314

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/87748
